dspy/teleprompt/finetune_teleprompter.py

```diff
@@ -0,0 +1,237 @@
+    a list of prompt-completion data dictionaries."""
+
+    prompt_completion_data = []
+    for data_dict in data:
+        trace = data_dict["trace"]
+        trace_prompt_comletion_data = build_messages_from_trace(
+            trace=trace, exclude_demos=exclude_demos,
+            try_to_record_lm_kwargs=try_to_record_lm_kwargs, program=program
+        )
+        for prompt_completion_dict in trace_prompt_comletion_data:
+            if keep_data_keys:
+                prompt_completion_dict = {**data_dict, **prompt_completion_dict}
+            prompt_completion_data.append(prompt_completion_dict)
+    return prompt_completion_data
+
+def build_messages_from_trace(
+        trace: List[Dict],
+        exclude_demos: bool=False,
+        try_to_record_lm_kwargs: bool = False,
+        program: Program = None,
+    ) -> Dict[str, List[Dict[str, Any]]]:
+    messages = []
+    # If the program is provided, build the predictor index to name mapping
+    if program:
+        pred_ind_to_name = {
+            ind: name for ind, (name, _) in enumerate(program.named_predictors())
+        }
+
+    # Build the prompt-completion data
+
+    adapter = dspy.settings.adapter or dspy.ChatAdapter()
+    data = []
+
+    # TODO: Make sure that this works for multi-stage pipelines
+    for pred_ind, (pred, inputs, outputs) in enumerate(trace):
+        # Get the demos from the predictor if exclude_demos is False
+        demos = [] if exclude_demos else pred.demos
+        messages = adapter.format(pred.signature, demos, inputs)
+        messages.append(adapter.format_turn(signature=pred.signature, values=outputs, role="assistant", incomplete=False))
+        data.append(messages)
+
+    return data
+
+def dummy_metric(example, pred, trace=None, frac=1.0):
+    return 1
+
+def bootstrap_data(
+        program: Program,
+        dataset: List[Example],
+        metric: Optional[Callable[
+            [Example, Prediction, Optional[List]], Union[bool, int, float]
+        ]] = dummy_metric,
+        num_threads = 1,
+        max_errors: int = 0
+    ) -> List[Dict[str, Any]]:
+    """Bootstrap example, prediction, trace, example_ind, score data for the program using the dataset."""
+    data = []
+
+    # Use Evaluate to call the program have the responses cached
+    cname = program.__class__.__name__
+    info = _INFO_BOOTSTRAP_DATA.format(len(dataset), cname, num_threads)
+    logger.info(info)
+    evaluator = Evaluate(
+        devset=dataset, num_threads=num_threads, display_progress=True, max_errors=max_errors, provide_traceback=True
+    )
+    evaluator(program, metric=metric)
+
+    data = []
+    for example in dataset:
+        data_dict = process_example(example, 0, program, metric)
+        if data_dict is not None:
+            data.append(data_dict)
+    
+    return data
+
+
+def process_example(example: Any, example_ind: int, program: Callable, metric: Optional[Callable] = None) -> Dict[str, Any]:
```

Comment:
I think this is called `bootstrap_one_example()` (or similar) in the `BootstrapFewshot`optimizer, maybe keep the consistency?

Reply:
Changed